client/app/pages/queries/components/QueryVisualizationTabs.jsx

```diff
@@ -82,6 +82,18 @@ const defaultVisualizations = [
   },
 ];
 
+
+const CachedVisualizationRenderer = React.memo(
+  function CachedVisRender(props) {
+    return <VisualizationRenderer {...props} />
+  },
+  (prevProps, nextProps) => (
+    prevProps.queryResult === nextProps.queryResult &&
+    prevProps.filters === nextProps.filters &&
+    prevProps.visualization === nextProps.visualization &&
+    prevProps.onFiltersChange === nextProps.onFiltersChange
```

Comment:
Have you figured out which prop caused component re-rendering?

Reply:
Inside `QuerySource`, the change of sql value will lead to 
```
setQuery(extend(query.clone(), { query: queryText }));
``` 

and `QueryVisualizationTabs`  component uses `query.visualizations`
```
<QueryVisualizationTabs   queryResult={queryResult}   visualizations={query.visualizations}
... />
```
I added the other checks just for safety.

Reply:
That doesn't explain the issue. If we change `query` object itself, but don't touch other fields - React will not update components which use that fields. So in this case `visualizations={query.visualizations}` cannot cause component re-rendering.

I'm just trying to find the root cause of the issue, because memoizing the whole component just fixes an aftereffect and may just hide some real problem.

Reply:
Make sense. I'll try to figure it out  as well! Thanks ~

Reply:
I make a demo app on [https://codepen.io/JerryJ/pen/yLJgYqO?editors=0011](https://codepen.io/JerryJ/pen/yLJgYqO?editors=0011)，I update obj outside and the child component still re-renders even the age property is never touched. You can check that.

Reply:
@kravets-levko doesn't the clone function recreate the visualizations, thus creating a new reference as well?

Reply:
What is triggering this is indeed probably any update in the Query Source page (most likely the Query one in this case). We've never memoized Visualizations, thus, any parent rerender will trigger a rerender on it:

<img width="1250" alt="Screen Shot 2020-10-23 at 11 10 09" src="https://user-images.githubusercontent.com/3356951/97014132-54285c00-1520-11eb-8d7d-25aa4fcb685b.png">

We currently have Visualization Widgets with `React.memo` for dashboards, but the visualizations themselves are not 😬. We do need to make sure this won't break anything.

Reply:
Maybe I can provide some tests, but I really don't know how many functions need to be tested. Any suggestions ?